### DISCUSSION TITLE
Change UI to scenes rather than code generated

### DISCUSSION THREAD

**Discussion by TokisanGames** [TokisanGames/Terrain3D#50] [Architecture]
It's very difficult for us to improve and build upon the UI when generated by code. It should be built as a scene in Godot and instanced by the code.

eg
```
const UI = "res://addons/gterrain/ui/GTerrainUI.tscn"

ui = load(UI).instantiate()
ui.ted = self
add_control_to_bottom_panel(ui, "Terrain Editor")
make_bottom_panel_item_visible(ui)
```

**Comment by outobugi** [comment]
I strongly disagree. I find it much easier to work on the UI when it's all in code.
The UI code could be separated to multiple scripts instead of sub classes.

**Reply by TokisanGames** [reply]
There are two of us now and eventually there will be more contributors who will find designing a UI with GUI tools much easier. Designing, experimenting and implementing the proper placement on something like this takes a long time. The discovery phase of something like this in code would take 2-10x longer.

![image](https://user-images.githubusercontent.com/632766/216732361-99a1e520-6368-4630-b2fc-a8348081f31d.png)

**Reply by outobugi** [reply]
10x longer? That's ridiculous.
The UI will not stay in GDScript. It will be moved to C++ and when that comes the transition is fast because the code and logic is already there.

**Reply by TokisanGames** [reply]
It's not ridiculous. I'm not talking about the speed of code. I'm talking about design time.

**Reply by outobugi** [reply]
> I'm not talking about the speed of code. I'm talking about design time.

I know.

**Reply by Calinou** [reply]
What I did for https://github.com/godotengine/godot/pull/79599 is to prototype the UI in Godot + GDScript with the editor, then rewrite it in C++.

To help yourself with rewriting the UI in C++ (and knowing which properties you should set in the constructor), I suggest looking at the saved `.tscn` file in a text editor. Only non-default properties are saved there.
